### PR TITLE
impl(GCS+gRPC): new options for gRPC metrics

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #13998

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14094)

### CODE REVIEW

File: google/cloud/storage/grpc_plugin.h [hunk 1]
```diff
@@ -60,6 +60,44 @@ struct [[deprecated(
  */
 google::cloud::storage::Client DefaultGrpcClient(Options opts = {});
 
+/**
+ * Enable gRPC telemetry for GCS RPCs.
+ *
+ * Troubleshooting problems with GCS over gRPC is difficult without some
+ * telemetry indicating how the client is configured, and what load balancing
+ * decisions.
```
Comment (dbolduc) [inline google/cloud/storage/grpc_plugin.h:68]: "decisions are made." ?

File: google/cloud/storage/grpc_plugin.h [hunk 2]
```diff
@@ -60,6 +60,44 @@ struct [[deprecated(
...
+ *
+ * When this option is enabled (the default) the GCS client will export the gRPC
```
Comment (dbolduc) [inline google/cloud/storage/grpc_plugin.h:70]: s/(the default)/(the default),/

File: google/cloud/storage/internal/grpc/default_options.cc [hunk 3]
```diff
@@ -31,6 +31,8 @@ namespace storage_internal {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace {
 
+auto constexpr kMinMetricsPeriod = std::chrono::seconds(60);
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/default_options.cc:34]: (just curious) Where did this value come from? The GCM minimum is a 5 second period.

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/grpc_plugin.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/grpc_plugin.h:68]: Fixed.

File: google/cloud/storage/grpc_plugin.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/grpc_plugin.h:70]: Fixed.

File: google/cloud/storage/internal/grpc/default_options.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc/default_options.cc]: I made it up, fixed.

File: google/cloud/storage/internal/grpc/default_options_test.cc [hunk 4]
```diff
@@ -132,11 +132,17 @@ TEST(DefaultOptionsGrpc, MetricsPeriod) {
...
 TEST(DefaultOptionsGrpc, MinMetricsPeriod) {
-  auto const options =
+  auto const o0 =
       DefaultOptionsGrpc(Options{}.set<storage_experimental::GrpcMetricsPeriod>(
-          std::chrono::seconds(1)));
-  EXPECT_GE(options.get<storage_experimental::GrpcMetricsPeriod>(),
-            std::chrono::seconds(60));
+          std::chrono::seconds(0)));
+  EXPECT_GT(o0.get<storage_experimental::GrpcMetricsPeriod>(),
+            std::chrono::seconds(0));
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/default_options_test.cc:139]: I think this should check for >= 5 seconds. Any less, and some of the exports will fail.

Unless you want to argue that "The GCM minimum may change, so we cannot say what it is, but we know it is greater than 0".

File: google/cloud/storage/internal/grpc/default_options_test.cc [hunk 5]
```diff
@@ -132,11 +132,17 @@ TEST(DefaultOptionsGrpc, MetricsPeriod) {
...
+
+  auto const m5 =
+      DefaultOptionsGrpc(Options{}.set<storage_experimental::GrpcMetricsPeriod>(
+          std::chrono::seconds(-5)));
+  EXPECT_GT(m5.get<storage_experimental::GrpcMetricsPeriod>(),
+            std::chrono::seconds(0));
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/default_options_test.cc:145]: same

File: google/cloud/storage/grpc_plugin.h [hunk 6]
```diff
@@ -60,6 +60,44 @@ struct [[deprecated(
...
+struct EnableGrpcMetrics {
+  using Type = bool;
+};
+
+/**
+ * gRPC telemetry export period.
+ *
+ * When `EnableGrpcMetrics` is enabled, this option controls the frequency at
+ * which metrics are exported to [Google Cloud Monitoring]. The default is 60
+ * seconds. Values below 60 seconds are ignored.
```
Comment (dbolduc) [inline google/cloud/storage/grpc_plugin.h:93]: s/60/5/

Comment (coryan) [reply google/cloud/storage/grpc_plugin.h:93]: https://github.com/googleapis/google-cloud-cpp/pull/14116

File: google/cloud/storage/internal/grpc/default_options_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc/default_options_test.cc:139]: Yes, I would like to argue that we don't care that much:

https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html

**Pull request merged**